### ISSUE TITLE
Peggo orchestrator uses only a single price oracle in IsBatchProfitable

### ISSUE THREAD

**Issue by facundomedica** [umee-network/peggo#181]
TOB-UMEE-033

The Peggo orchestrator relays ethereum batches only if they are profitable (figure 33.1). Profitability is determined by using the ETH/USD price fetched from a single source – the CoinGecko API. This creates a single point of failure for the Peggo relayer as if the used API gets hacked, an attacker will be able to choose whether to relay the batches.

----

## For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by RafilxTenfen** [comment]
__Ideas__
1. Use the [umee price feeder](https://github.com/umee-network/umee/tree/main/price-feeder) as a package to grab the price of a pair 
   - Using it as a package we are going to have multiples providers, but probably we would duplicate some prices that are already on-chain  
1. Query the pair price on chain
   - Querying the prices on chain would implicate that we had a lot of prices needed just for peggo 
1. Try to query it on chain, if the pair is not there, use the price feeder as a package 
   
- The drawback is that peggo doesn't know previously which token pair is needed. So, we would need to implement a way to add new pairs after the price-feeder is created  

What do you guys think? @facundomedica @alexanderbez @adamewozniak @toteki

**Comment by alexanderbez** [comment]
@RafilxTenfen how would one use the price-feeder as a package/library? IIRC, we expose the prices via an API. It would make more sense to to query the API instead. But at that point, we might as well query on chain for canonical prices.

So I like (2) the most. However, I know there is some weird interplay between getting native and ERC20 assets in Peggo. I'd have to defer to @facundomedica here :)

**Comment by RafilxTenfen** [comment]
My idea to use the price-feeder as a package/library was that we would need to make a refractory to separate the logic to get and compute the prices of the providers from the vote/prevote and tick logic and then the orchestrator would create his own price-feeder which computes the volume-weighted average price from 5 providers

**Comment by toteki** [comment]
How many pairs do we need for this? Issue only mentions ETH/USD but are there others involved?

**Comment by RafilxTenfen** [comment]
It needs the pairs of all tokens that are crossing in the bridge [TOKEN/USD](https://github.com/umee-network/peggo/blob/c3546a4d642ee6c2cbcb4e013d771b8a9633a4e4/orchestrator/relayer/batch_relaying.go#L250) as far as I understood

**Comment by alexanderbez** [comment]
> How many pairs do we need for this? Issue only mentions ETH/USD but are there others involved?

For all pairs that can be bridged. AFAIK, the price-feeder only acts as an oracle for assets supported in the leverage module's registry.

So using the price-feeder won't be sufficient I think.

**Comment by facundomedica** [comment]
Yeah, so the bridge can send any token you throw at it, so using the oracle prices isn't really an option.

Also, we "only" have the token address (we also have the symbol but that must never be used as someone else can just create an identical one) so we need to be able to query prices using the address. For the tokens deployed by us we have a constants list that matches each one with Coingecko's coin ID.

Gravity Bridge's relayer is using Uniswap at the moment, btw.

**Comment by facundomedica** [comment]
To sum up our call:
- Use pf's providers
- Add the option to lazily start a price pair

If we add the "lazy start" I don't think we need the token whitelist for now

**Comment by RafilxTenfen** [comment]
As discussed earlier with @facundomedica 

- To solve this issue we could get the symbol from the CoinGecko [API](https://www.coingecko.com/en/api/documentation) using the contract address 
```shell
$~ curl -X 'GET' \
  'https://api.coingecko.com/api/v3/coins/ethereum/contract/0x514910771af9ca656af840dff83e8264ecf986ca' \
  -H 'accept: application/json'
``` 
- Keep that information saved in memory and get the price with the [umee provider package](https://github.com/umee-network/umee/tree/main/price-feeder/oracle/provider) using the SYMBOL/USDC and compute all the prices received with [tvwap](https://github.com/umee-network/umee/issues/542)

- To make this happen, one function in the providers need to be added/public 
```go
// SubscribeTickers subscribe to all currency pairs and
// add the new ones into the provider subscribed pairs.
func (p *provider) SubscribeTickers(cps ...types.CurrencyPair) error {
....
}
```
- Also implement the use of providers in [peggo batch relaying](https://github.com/umee-network/peggo/blob/c3546a4d642ee6c2cbcb4e013d771b8a9633a4e4/orchestrator/relayer/batch_relaying.go#L250) to check whether a batch is profitable to be bridge

cc @alexanderbez

**Comment by alexanderbez** [comment]
@RafilxTenfen and what does this get us? Aren't we using coingecko directly already? Sorry, I'm just trying to piece it all together in my head haha

**Comment by RafilxTenfen** [comment]
@alexanderbez Yes, we are using coingecko to get the price, so it is the only provider

- In this new implementation, we are going to use coingecko API just to get the correct symbol, the price itself is going to use multiple providers 

- At least this was our idea to solve the problem of getting the correct symbol of a contract address, another idea is to keep a map of tokens hardcoded  with their expected symbol

**Comment by alexanderbez** [comment]
Will this work @facundomedica? I thought you said there would be an issue with token contracts?

**Comment by facundomedica** [comment]
@alexanderbez yes sir! So the issue we have now is that we don't have a way to turn a contract address into a symbol to later use it in the price feeder as a ticker. What Rafael suggests is the fix we need imo

**Comment by alexanderbez** [comment]
Perfect! Let's do it!

But one thing I'd like to point out and this just came to me, is that we won't be using the price-feeder at some point in the future. If/when we switch to using Chainlink, I'm not sure we'll have a need for the price-feeder, but we can still keep it around.

**Comment by RafilxTenfen** [comment]
@alexanderbez if you think that we should use chainlink as a price-feeder I could implement directly the chainlink adapter instead of using our price-feeder

**Comment by alexanderbez** [comment]
Chainlink is down the road @RafilxTenfen -- we'll need to figure out how to use Chainlink in the future. For now, let's stick with what we have. Because using Chainlink will change our architecture completely.
